### ISSUE TITLE
Research feasibility of using GitHub Actions for executing integration testing

### ISSUE THREAD

**Issue by jakerobb** [k8ssandra/k8ssandra#104]
We want to add some more CI/CD integration, but for several things we want to add, we'll need more than the 2cpu/7gb limits available to us on GitHub actions.

@jsanda mentioned that "self-hosted runners" are a thing -- we might be able to host something on our internal infrastructure that can take jobs from GitHub and publish results back afterward.

This issue encompasses the research effort to determine what is possible and whether it helps us.

**Comment by jdonenine** [comment]
I think the overall goal here is to come to a CI/CD approach that is functional, sufficient, and approachable from the community.  We want to make sure we've got the guards in place to let folks get involved in the project - that means having runs both automated and visible to them.

**Comment by jakerobb** [comment]
When we say "visible," does that mean they need to be able to monitor builds _as they run_? Or would it be sufficient if they could only see build _results_?

**Comment by jsanda** [comment]
We definitely want to make build results visible. It would be nice to also allow users to monitor builds as they run, assuming monitoring in this context means to observe.

**Comment by jdonenine** [comment]
I'm hoping to generalize this issue a bit from its original description to evaluating GH Actions vs CircleCI for integration testing - basically, can what @adejanovski has prototyped in CircleCI be brought to the GH world - particularly the ability for someone working on a fork of the project to run the integration tests on their own free tier of actions.

**Comment by adejanovski** [comment]
After investigating why my GHA runs were failing on the Medusa S3 test, it turns out I had a bad secret stored for the S3 credentials.
I was able to [run the current set of integration tests](https://github.com/adejanovski/k8ssandra/actions/runs/624102339) faster than with CircleCI:
CircleCI run:
![Capture d’écran 2021-03-05 à 11 55 21](https://user-images.githubusercontent.com/5096002/110106896-84bfb280-7daa-11eb-9dc2-043778b736af.png)

GHA run:
![Capture d’écran 2021-03-05 à 11 55 36](https://user-images.githubusercontent.com/5096002/110106906-88ebd000-7daa-11eb-916d-08b2b9002d8a.png)

I also found [an action](https://github.com/actions/upload-artifact) that will allow us to store test artifacts for deeper investigation on failing runs.

While CircleCI gives us 4 concurrent runners per fork, GHA will give us 20 concurrent runners per fork, which will largely reduce our feedback loop.
One feature that we'll be missing from CircleCI though is the ability to SSH into the CI runners, which is a convenient one to debug CI runs.

Aside from this, I'm now +1 on using GHA for integration testing as there are no technical blockers anymore.

**Comment by adejanovski** [comment]
@jdonenine, let me know if the above is enough for you to close the issue and open a new ticket for the implementation.

**Comment by adejanovski** [comment]
Here's another 3rd party action that allows us to setup SSH debugging in GHA in case of workflow failure: https://github.com/marketplace/actions/debugging-with-tmate

**Comment by adejanovski** [comment]
I've implemented the collection of artifacts in case of failure to help investigating issues:
![Fix-artifacts-extraction-·-adejanovski-k8ssandra-591b7d0](https://user-images.githubusercontent.com/5096002/110946558-4d6b7b80-833f-11eb-9020-626f34400f8d.png)

The extraction takes place [in this script](https://github.com/adejanovski/k8ssandra/blob/alex/add-circleci-build/.github/scripts/collect_artifacts.sh).
It runs some `kubectl get/describe/logs` commands and stores the outputs in files that will be compressed and attached to the workflow run.

I've also added a step in the workflow (which is commented by default) to enable ssh access to the runner in case of failure. It helps investigating complex errors and should, IMO, not be enabled by default. One of the reasons is that it will prevent the workflow run from finishing until it reaches the workflow timeout.
